framework/src/transfers/MultiAppInterpolationTransfer.C

```diff
@@ -66,492 +83,498 @@ MultiAppInterpolationTransfer::MultiAppInterpolationTransfer(const InputParamete
+  std::vector<Number> & src_vals(idi->get_source_vals());
+
+  // How much we want to translate mesh if users ask
+  std::unordered_map<dof_id_type, Point> from_tranforms;
+  std::set<subdomain_id_type> exclude_block_ids;
+  if (_shrink_gap_width > 0 && _shrink_mesh == "source")
+  {
+    computeTransformation(*from_moose_mesh, from_tranforms);
+    auto exclude_subdomainids = from_moose_mesh->getSubdomainIDs(_exclude_gap_blocks);
+    exclude_block_ids.insert(exclude_subdomainids.begin(), exclude_subdomainids.end());
+  }
 
-      // EquationSystems & from_es = from_sys.get_equation_systems();
+  // The solution from the system with which the from_var is associated
+  NumericVector<Number> & from_solution = *from_sys.solution;
 
-      NumericVector<Number> & from_solution = *from_sys.solution;
+  std::set<subdomain_id_type> subdomainids;
+  std::vector<subdomain_id_type> include_block_ids;
+  if (from_is_nodal)
+  {
+    for (const auto & from_node : from_mesh->local_node_ptr_range())
+    {
+      // Assuming LAGRANGE!
+      if (from_node->n_comp(from_sys_num, from_var_num) == 0)
+        continue;
 
-      InverseDistanceInterpolation<LIBMESH_DIM> * idi;
+      Point translate(0);
 
-      switch (_interp_type)
+      if (from_tranforms.size() > 0)
       {
-        case 0:
-          idi = new InverseDistanceInterpolation<LIBMESH_DIM>(from_sys.comm(), _num_points, _power);
-          break;
-        case 1:
-          idi = new RadialBasisInterpolation<LIBMESH_DIM>(from_sys.comm(), _radius);
-          break;
-        default:
-          mooseError("Unknown interpolation type!");
+        subdomainIDsNode(*from_moose_mesh, *from_node, subdomainids);
+        // Check if node is excluded
+        // Node will be excluded if it is in the interior of excluded subdomains
+        include_block_ids.clear();
+        include_block_ids.resize(std::max(subdomainids.size(), exclude_block_ids.size()));
+        auto it = std::set_difference(subdomainids.begin(),
+                                      subdomainids.end(),
+                                      exclude_block_ids.begin(),
+                                      exclude_block_ids.end(),
+                                      include_block_ids.begin());
+
+        include_block_ids.resize(it - include_block_ids.begin());
+        // Node is not excluded
+        if (include_block_ids.size())
+          translate = from_tranforms[*include_block_ids.begin()];
+        else
+          continue;
       }
 
-      std::vector<Point> & src_pts(idi->get_source_points());
-      std::vector<Number> & src_vals(idi->get_source_vals());
+      // Push value and point to KDTree
+      dof_id_type from_dof = from_node->dof_number(from_sys_num, from_var_num, 0);
+      src_vals.push_back(from_solution(from_dof));
+      src_pts.push_back(*from_node + translate + from_app_position);
+    }
+  }
+  else
+  {
+    std::vector<Point> points;
+    for (const auto & from_elem :
```

Comment:
This is an idiom that is seen throughout MOOSE that is misleading. What this actually ends up creating is:
`Elem * const &`, e.g. the pointer is `const` but the `Elem` is not. I prefer this:
```c++
for (const auto * from_elem
```
or if you want to be really paranoid:
```c++
for (const auto * const elem
```
which would ensure that both the pointer and the pointed-to object are `const`.
This ensures that `Elem` is `const` which is probably what you really wanted.

You can try out this test program:
```c++
#include <iostream>
#include <vector>

class A {
public:
  void foo() const { std::cout << "Hi there!\n"; }

  void bar() { std::cout << "Hi there!\n"; }
};

int main() {
  std::vector<A *> a(5);

  for (auto *elem : a)
    elem = new A;

  for (const auto *const elem : a) {
    elem->foo();

    // The line below won't compile because we are trying to call a non-const
    // method with a const object

    // elem->bar();
  }

  for (const auto &elem : a) {
    elem->foo();

    // This line compiles fine however, because we are calling through pointer
    // to non-const. Not as safe as we might have thought!
    elem->bar();
  }

  for (auto *elem : a)
    delete elem;
}
```

Reply:
It is very interesting.   A good lesson!